src/daft-csv/src/read.rs

```diff
@@ -7,325 +7,293 @@ use arrow2::{
+    read_options: Option<CsvReadOptions>,
     io_client: Arc<IOClient>,
     io_stats: Option<IOStatsRef>,
     multithreaded_io: bool,
-    schema: Option<SchemaRef>,
-    buffer_size: Option<usize>,
-    chunk_size: Option<usize>,
     max_chunks_in_flight: Option<usize>,
 ) -> DaftResult<Table> {
     let runtime_handle = get_runtime(multithreaded_io)?;
     let _rt_guard = runtime_handle.enter();
     runtime_handle.block_on(async {
-        read_csv_single(
+        read_csv_single_into_table(
             uri,
-            column_names,
-            include_columns,
-            num_rows,
-            has_header,
-            delimiter.unwrap_or(b','),
-            double_quote,
+            convert_options,
+            parse_options,
+            read_options,
             io_client,
             io_stats,
-            schema,
-            buffer_size,
-            chunk_size,
             max_chunks_in_flight,
         )
         .await
     })
 }
 
 #[allow(clippy::too_many_arguments)]
-async fn read_csv_single(
+pub fn read_csv_bulk(
+    uris: &[&str],
+    convert_options: Option<CsvConvertOptions>,
+    parse_options: Option<CsvParseOptions>,
+    read_options: Option<CsvReadOptions>,
+    io_client: Arc<IOClient>,
+    io_stats: Option<IOStatsRef>,
+    multithreaded_io: bool,
+    max_chunks_in_flight: Option<usize>,
+    num_parallel_tasks: usize,
+) -> DaftResult<Vec<Table>> {
+    let runtime_handle = get_runtime(multithreaded_io)?;
+    let _rt_guard = runtime_handle.enter();
+    let tables = runtime_handle
+        .block_on(async move {
+            // Launch a read task per URI, throttling the number of concurrent file reads to num_parallel tasks.
+            let task_stream = futures::stream::iter(uris.iter().enumerate().map(|(i, uri)| {
+                let (uri, convert_options, parse_options, read_options, io_client, io_stats) = (
+                    uri.to_string(),
+                    convert_options.clone(),
+                    parse_options.clone(),
+                    read_options.clone(),
+                    io_client.clone(),
+                    io_stats.clone(),
+                );
+                tokio::task::spawn(async move {
+                    let table = read_csv_single_into_table(
+                        uri.as_str(),
+                        convert_options,
+                        parse_options,
+                        read_options,
+                        io_client,
+                        io_stats,
+                        max_chunks_in_flight,
+                    )
+                    .await?;
+                    Ok((i, table))
+                })
+            }));
+            task_stream
+                // Each task is annotated with its position in the output, so we can use unordered buffering to help mitigate stragglers
+                // and sort the task results at the end.
+                .buffer_unordered(num_parallel_tasks)
+                .try_collect::<Vec<_>>()
```

Comment:
You can terminate early here once you hit the `limit` rows.

Reply:
Ah yes I have that as a follow-up, would you like for me to add early termination in this PR? Should be easy to add.